core/base/mergeTreeClustering/PathMappingDistance.h

```diff
@@ -0,0 +1,360 @@
+      dataType d2 = nodes2[p2];
+      dataType d = std::abs(std::abs(b1 - d1) - std::abs(b2 - d2));
+      return squared ? d * d : d;
+    }
+
+  public:
+    PathMappingDistance() {
+      this->setDebugMsgPrefix(
+        "MergeTreeDistance"); // inherited from Debug: prefix will be printed at
+                              // the beginning of every msg
+    }
+    ~PathMappingDistance() override = default;
+
+    void setBaseMetric(int m) {
+      baseMetric = m;
+    }
+
+    void setAssignmentSolver(int assignmentSolver) {
+      assignmentSolverID_ = assignmentSolver;
+    }
+
+    void setSquared(bool s) {
+      squared = s;
+    }
+
+    template <class dataType>
+    dataType editDistance_path(std::vector<dataType> &nodes1,
+                               std::vector<std::vector<int>> &topo1,
+                               int rootID1,
+                               std::vector<dataType> &nodes2,
+                               std::vector<std::vector<int>> &topo2,
+                               int rootID2) {
+
+      // initialize memoization tables
+
+      std::vector<std::vector<int>> predecessors1(nodes1.size());
+      std::vector<std::vector<int>> predecessors2(nodes2.size());
+
+      int depth1 = 0;
+      int depth2 = 0;
+      std::stack<int> stack;
+      stack.push(rootID1);
+      while(!stack.empty()) {
+        int nIdx = stack.top();
+        stack.pop();
+        depth1 = std::max((int)predecessors1[nIdx].size(), depth1);
+        for(int cIdx : topo1[nIdx]) {
+          stack.push(cIdx);
+          predecessors1[cIdx].reserve(predecessors1[nIdx].size() + 1);
+          predecessors1[cIdx].insert(predecessors1[cIdx].end(),
+                                     predecessors1[nIdx].begin(),
+                                     predecessors1[nIdx].end());
+          predecessors1[cIdx].push_back(nIdx);
+        }
+      }
+      stack.push(rootID2);
+      while(!stack.empty()) {
+        int nIdx = stack.top();
+        stack.pop();
+        depth2 = std::max((int)predecessors2[nIdx].size(), depth2);
+        for(int cIdx : topo2[nIdx]) {
+          stack.push(cIdx);
+          predecessors2[cIdx].reserve(predecessors2[nIdx].size() + 1);
+          predecessors2[cIdx].insert(predecessors2[cIdx].end(),
+                                     predecessors2[nIdx].begin(),
+                                     predecessors2[nIdx].end());
+          predecessors2[cIdx].push_back(nIdx);
+        }
+      }
+
+      size_t nn1 = nodes1.size();
+      size_t nn2 = nodes2.size();
+      size_t dim1 = 1;
+      size_t dim2 = (nn1 + 1) * dim1;
+      size_t dim3 = (depth1 + 1) * dim2;
+      size_t dim4 = (nn2 + 1) * dim3;
+      // size_t dim1_ = 1;
+      // size_t dim2_ = (nn1 + 1) * dim1;
+      // size_t dim3_ = (nn1 + 1) * dim2;
+      // size_t dim4_ = (nn2 + 1) * dim3;
+
+      std::unique_ptr<dataType[]> memT(
+        new dataType[(nn1 + 1) * (depth1 + 1) * (nn2 + 1) * (depth2 + 1)]);
```

Comment:
please re-factor to avoid the usage of any `new` and `delete`.

Reply:
I agree. Why not use the commented `std::vector` declaration below instead?

Reply:
yeah I'm afraid there might be some memory leaks there.
I'd recommend to switch to the `std::vector`